File: examples/gem2md.zig
//! This example implements a AST based parsing which
//! first reads in the whole file into a document, then
//! renders the parsed document again.

const std = @import("std");
const gemtext = @import("gemtext");

pub fn main() !void {
    var document = try gemtext.Document.parse(std.heap.page_allocator, std.io.getStdIn().reader());
    defer document.deinit();

    try gemtext.renderer.markdown(document.fragments.items, std.io.getStdOut().writer());
}

File: examples/gem2md.c
//! This example implements a AST based parsing which
//! first reads in the whole file into a document, then
//! renders the parsed document again.

#include <stdio.h>
#include <stdbool.h>
#include "gemtext.h"

void renderToStream(void *ctx, char const *buffer, size_t length)
{
  size_t offset = 0;
  while (offset < length)
  {
    int len = fwrite(buffer + offset, 1, length, (FILE *)ctx);
    if (len < 0)
      return;
    if (len == 0)
      break;
    offset += (size_t)len;
  }
}

int main()
{
  struct gemtext_document document;

  if (gemtextDocumentParseFile(&document, stdin) != GEMTEXT_SUCCESS)
    return 1;

  enum gemtext_error error = gemtextRender(
      GEMTEXT_RENDER_MARKDOWN,
      document.fragments,
      document.fragment_count,
      stdout,
      renderToStream);

  gemtextDocumentDestroy(&document);

  if (error != GEMTEXT_SUCCESS)
    return 1;

  return 0;
}